In [35]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.ensemble import RandomForestClassifier,RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV,RandomizedSearchCV

from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

from sklearn.metrics import r2_score,mean_absolute_error,mean_squared_error

import warnings
warnings.filterwarnings("ignore")

from sklearn.preprocessing import LabelEncoder

In [3]:
df = pd.read_csv("Hotel Reservations.csv")
df

Booking_ID  no_of_adults  no_of_children  no_of_weekend_nights  \
0       INN00001             2               0                     1   
1       INN00002             2               0                     2   
2       INN00003             1               0                     2   
3       INN00004             2               0                     0   
4       INN00005             2               0                     1   
...          ...           ...             ...                   ...   
36270   INN36271             3               0                     2   
36271   INN36272             2               0                     1   
36272   INN36273             2               0                     2   
36273   INN36274             2               0                     0   
36274   INN36275             2               0                     1   

       no_of_week_nights type_of_meal_plan  required_car_parking_space  \
0                      2       Meal Plan 1                           0   
1                      3      Not Selected                           0   
2                      1       Meal Plan 1                           0   
3                      2       Meal Plan 1                           0   
4                      1      Not Selected                           0   
...                  ...               ...                         ...   
36270                  6       Meal Plan 1                           0   
36271                  3       Meal Plan 1                           0   
36272                  6       Meal Plan 1                           0   
36273                  3      Not Selected                           0   
36274                  2       Meal Plan 1                           0   

      room_type_reserved  lead_time  arrival_year  arrival_month  \
0            Room_Type 1        224          2017             10   
1            Room_Type 1          5          2018             11   
2            Room_Type 1          1          2018              2   
3            Room_Type 1        211          2018              5   
4            Room_Type 1         48          2018              4   
...                  ...        ...           ...            ...   
36270        Room_Type 4         85          2018              8   
36271        Room_Type 1        228          2018             10   
36272        Room_Type 1        148          2018              7   
36273        Room_Type 1         63          2018              4   
36274        Room_Type 1        207          2018             12   

       arrival_date market_segment_type  repeated_guest  \
0                 2             Offline               0   
1                 6              Online               0   
2                28              Online               0   
3                20              Online               0   
4                11              Online               0   
...             ...                 ...             ...   
36270             3              Online               0   
36271            17              Online               0   
36272             1              Online               0   
36273            21              Online               0   
36274            30             Offline               0   

       no_of_previous_cancellations  no_of_previous_bookings_not_canceled  \
0                                 0                                     0   
1                                 0                                     0   
2                                 0                                     0   
3                                 0                                     0   
4                                 0                                     0   
...                             ...                                   ...   
36270                             0                                     0   
36271                             0                                     0   
36272                            

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 19 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   Booking_ID                            36275 non-null  object 
 1   no_of_adults                          36275 non-null  int64  
 2   no_of_children                        36275 non-null  int64  
 3   no_of_weekend_nights                  36275 non-null  int64  
 4   no_of_week_nights                     36275 non-null  int64  
 5   type_of_meal_plan                     36275 non-null  object 
 6   required_car_parking_space            36275 non-null  int64  
 7   room_type_reserved                    36275 non-null  object 
 8   lead_time                             36275 non-null  int64  
 9   arrival_year                          36275 non-null  int64  
 10  arrival_month                         36275 non-null  int64  
 11  arrival_date   

In [5]:
df.drop(['Booking_ID','type_of_meal_plan','market_segment_type'],axis =1,inplace = True)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 16 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   no_of_adults                          36275 non-null  int64  
 1   no_of_children                        36275 non-null  int64  
 2   no_of_weekend_nights                  36275 non-null  int64  
 3   no_of_week_nights                     36275 non-null  int64  
 4   required_car_parking_space            36275 non-null  int64  
 5   room_type_reserved                    36275 non-null  object 
 6   lead_time                             36275 non-null  int64  
 7   arrival_year                          36275 non-null  int64  
 8   arrival_month                         36275 non-null  int64  
 9   arrival_date                          36275 non-null  int64  
 10  repeated_guest                        36275 non-null  int64  
 11  no_of_previous_

In [9]:
lblEnc = LabelEncoder()
df["room_type_reserved"] = lblEnc.fit_transform(df["room_type_reserved"])

In [10]:
df

no_of_adults  no_of_children  no_of_weekend_nights  no_of_week_nights  \
0                 2               0                     1                  2   
1                 2               0                     2                  3   
2                 1               0                     2                  1   
3                 2               0                     0                  2   
4                 2               0                     1                  1   
...             ...             ...                   ...                ...   
36270             3               0                     2                  6   
36271             2               0                     1                  3   
36272             2               0                     2                  6   
36273             2               0                     0                  3   
36274             2               0                     1                  2   

       required_car_parking_space  room_type_reserved  lead_time  \
0                               0                   0        224   
1                               0                   0          5   
2                               0                   0          1   
3                               0                   0        211   
4                               0                   0         48   
...                           ...                 ...        ...   
36270                           0                   3         85   
36271                           0                   0        228   
36272                           0                   0        148   
36273                           0                   0         63   
36274                           0                   0        207   

       arrival_year  arrival_month  arrival_date  repeated_guest  \
0              2017             10             2               0   
1              2018             11             6               0   
2              2018              2            28               0   
3              2018              5            20               0   
4              2018              4            11               0   
...             ...            ...           ...             ...   
36270          2018              8             3               0   
36271          2018             10            17               0   
36272          2018              7             1               0   
36273          2018              4            21               0   
36274          2018             12            30               0   

       no_of_previous_cancellations  no_of_previous_bookings_not_canceled  \
0                                 0                                     0   
1                                 0                                     0   
2                                 0                                     0   
3                                 0                                     0   
4                                 0                                     0   
...                             ...                                   ...   
36270                             0                                     0   
36271                             0                                     0   
36272                             0                                     0   
36273                             0                                     0   
36274                             0                                     0   

       avg_price_per_room  no_of_special_requests booking_status  
0                   65.00                       0   Not_Canceled  
1                  106.68                       1   Not_Canceled  
2                   60.00                       0       Canceled  
3                  100.00                       0       Canceled  
4                   94.50                       0       Canceled  
...                   ...                     ...            ...  
36270             

In [12]:
df["booking_status"].replace({"Canceled":0,"Not_Canceled":1},inplace = True)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 16 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   no_of_adults                          36275 non-null  int64  
 1   no_of_children                        36275 non-null  int64  
 2   no_of_weekend_nights                  36275 non-null  int64  
 3   no_of_week_nights                     36275 non-null  int64  
 4   required_car_parking_space            36275 non-null  int64  
 5   room_type_reserved                    36275 non-null  int32  
 6   lead_time                             36275 non-null  int64  
 7   arrival_year                          36275 non-null  int64  
 8   arrival_month                         36275 non-null  int64  
 9   arrival_date                          36275 non-null  int64  
 10  repeated_guest                        36275 non-null  int64  
 11  no_of_previous_

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36275 entries, 0 to 36274
Data columns (total 16 columns):
 #   Column                                Non-Null Count  Dtype  
---  ------                                --------------  -----  
 0   no_of_adults                          36275 non-null  int64  
 1   no_of_children                        36275 non-null  int64  
 2   no_of_weekend_nights                  36275 non-null  int64  
 3   no_of_week_nights                     36275 non-null  int64  
 4   required_car_parking_space            36275 non-null  int64  
 5   room_type_reserved                    36275 non-null  int32  
 6   lead_time                             36275 non-null  int64  
 7   arrival_year                          36275 non-null  int64  
 8   arrival_month                         36275 non-null  int64  
 9   arrival_date                          36275 non-null  int64  
 10  repeated_guest                        36275 non-null  int64  
 11  no_of_previous_

In [15]:
# Model Evaluation

In [16]:
x = df.drop("booking_status",axis=1)
y = df["booking_status"]

In [18]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size =0.2,
                                random_state=1,stratify=y)

In [20]:
#model Trainig
rf_model = RandomForestClassifier()
rf_model.fit(x_train,y_train)

RandomForestClassifier()

In [21]:
#Testing Data Accuracy
y_pred = rf_model.predict(x_test)

acc_test = accuracy_score(y_test,y_pred)
print("Accuracy of Data Testing = ",acc_test)
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy of Data Testing =  0.8891798759476224
[[1871  506]
 [ 298 4580]]
              precision    recall  f1-score   support

           0       0.86      0.79      0.82      2377
           1       0.90      0.94      0.92      4878

    accuracy                           0.89      7255
   macro avg       0.88      0.86      0.87      7255
weighted avg       0.89      0.89      0.89      7255



In [22]:
# Training Data Accuracy
y_pred = rf_model.predict(x_train)

acc_train = accuracy_score(y_train,y_pred)
print("Accuracy of Training Data = ",acc_train)
print(confusion_matrix(y_train,y_pred))
print(classification_report(y_train,y_pred))

Accuracy of Training Data =  0.9936940041350792
[[ 9396   112]
 [   71 19441]]
              precision    recall  f1-score   support

           0       0.99      0.99      0.99      9508
           1       0.99      1.00      1.00     19512

    accuracy                           0.99     29020
   macro avg       0.99      0.99      0.99     29020
weighted avg       0.99      0.99      0.99     29020



In [25]:
# HyperParameters 
# 1)Randomised Search Cv
hype = {"criterion" :["gini","entropy"],
       "max_depth":np.arange(1,10),
        "n_estimators":np.arange(10,100),
        "min_samples_split":np.arange(2,10),
        "min_samples_leaf":np.arange(1,10)
       
       }

rf_clf = RandomForestClassifier()

In [26]:
rscv_clf = RandomizedSearchCV(rf_clf,hype,cv=5)
rscv_clf.fit(x_train,y_train)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(),
                   param_distributions={'criterion': ['gini', 'entropy'],
                                        'max_depth': array([1, 2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_leaf': array([1, 2, 3, 4, 5, 6, 7, 8, 9]),
                                        'min_samples_split': array([2, 3, 4, 5, 6, 7, 8, 9]),
                                        'n_estimators': array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
       27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43,
       44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60,
       61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77,
       78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94,
       95, 96, 97, 98, 99])})

In [27]:
rscv_clf.best_params_

{'n_estimators': 97,
 'min_samples_split': 4,
 'min_samples_leaf': 1,
 'max_depth': 8,
 'criterion': 'gini'}

In [28]:
rf_model = RandomForestClassifier(n_estimators=97,min_samples_leaf=1,
                                 min_samples_split=4,max_depth=8,criterion='gini')
rf_model.fit(x_train,y_train)

RandomForestClassifier(max_depth=8, min_samples_split=4, n_estimators=97)

In [29]:
#Testing Data Accuracy
y_pred = rf_model.predict(x_test)

acc_test = accuracy_score(y_test,y_pred)
print("Accuracy of Data Testing = ",acc_test)
print(confusion_matrix(y_test,y_pred))
print(classification_report(y_test,y_pred))

Accuracy of Data Testing =  0.8379048931771192
[[1383  994]
 [ 182 4696]]
              precision    recall  f1-score   support

           0       0.88      0.58      0.70      2377
           1       0.83      0.96      0.89      4878

    accuracy                           0.84      7255
   macro avg       0.85      0.77      0.80      7255
weighted avg       0.84      0.84      0.83      7255



In [30]:
# Training Data Accuracy
y_pred = rf_model.predict(x_train)

acc_train = accuracy_score(y_train,y_pred)
print("Accuracy of Training Data = ",acc_train)
print(confusion_matrix(y_train,y_pred))
print(classification_report(y_train,y_pred))

Accuracy of Training Data =  0.8459682977257064
[[ 5652  3856]
 [  614 18898]]
              precision    recall  f1-score   support

           0       0.90      0.59      0.72      9508
           1       0.83      0.97      0.89     19512

    accuracy                           0.85     29020
   macro avg       0.87      0.78      0.81     29020
weighted avg       0.85      0.85      0.84     29020



In [31]:
# Testing Accuracy  = 0.88
# Trtrainig Accuarcy =0.99

# Using Hyperparameter
# Testing Accuracy  =  0.83
# Trtrainig Accuarcy =0.84

# Regression

In [33]:
x = df.drop("avg_price_per_room",axis =1)
y = df["avg_price_per_room"]

In [37]:
# splitting
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=1)

In [38]:
# Model Training
rf_model = RandomForestRegressor()
rf_model.fit(x_train,y_train)

RandomForestRegressor()

In [39]:
# Testing Data Accuracy
y_pred = rf_model.predict(x_test)

r2 = r2_score(y_test,y_pred)
print("Accuracy Of Testing = ",r2)
print()

mae = mean_absolute_error(y_test,y_pred)
print("Mean absolute Error = ",mae)
print()

mse = mean_squared_error(y_test,y_pred)
print("Mean Squared Error = ",mse)
print()

rmse =np.sqrt(mse)
print("Root Mean Squared = ",rmse)



Accuracy Of Testing =  0.7022644385827171

Mean absolute Error =  10.604842864750166

Mean Squared Error =  359.6184046575885

Root Mean Squared =  18.963607374589586


In [40]:
# Traing Data Accuracy
y_pred = rf_model.predict(x_train)

r2 = r2_score(y_train,y_pred)
print("Accuracy Of Testing = ",r2)
print()

mae = mean_absolute_error(y_train,y_pred)
print("Mean absolute Error = ",mae)
print()

mse = mean_squared_error(y_train,y_pred)
print("Mean Squared Error = ",mse)
print()

rmse =np.sqrt(mse)
print("Root Mean Squared = ",rmse)



Accuracy Of Testing =  0.9501917257542616

Mean absolute Error =  4.193952645489162

Mean Squared Error =  61.61557971931228

Root Mean Squared =  7.849559205414804


# Hyperparameter
1)Grid Search Cv

In [ ]:
rf_model = RandomForestRegressor()
gscv_reg = GridSearchCV(rf_model,hype,cv= 5)
gscv_reg.fit(x_train,y_train)

In [ ]:
gscv_reg.best_params_

In [ ]:
rf_model = RandomForestRegressor(n_estimators = ,
                                criterion = ,
                                max_depth = ,
                                min_samples_leaf = ,
                                min_samples_split = 
                                )
rf_model.fit(x_train,y_train)

In [ ]:
y_pred = rf_model.predict(x_test)
# Testing Data Accuracy 
r2 = r2_score(y_test,y_pred)
print("Accuracy of Testing is = ",r2)
print()
mse= mean_squared_error(y_test,y_train)
print("Mean Squared Error is = ",mae)
print()
rmse =np.sqrt(mse)
print("Root Mean Squared Error = ",rmse)
print()
mae = mean_absolute_error(y_test,y_pred)
print("Mean absolute Error = ",mae)



In [ ]:
# Training Data Accuracy
y_pred = rf_model.predict(x_train)

r2 = r2_score(y_train,y_pred)
print("Accuracy Of Testing = ",r2)
print()

mae = mean_absolute_error(y_train,y_pred)
print("Mean absolute Error = ",mae)
print()

mse = mean_squared_error(y_train,y_pred)
print("Mean Squared Error = ",mse)
print()

rmse =np.sqrt(mse)
print("Root Mean Squared = ",rmse)



#  In Regression Concern 
 
 ### Without parameters

## Testing accuracy = 0.70
## Training Accurcy = 0.95

# Hyperameter using GridSearchCV

## Testing accuracy =
## Training accuracy = 

